pkg/liqo-controller-manager/nodefailure-controller/nodefailure_controller.go

```diff
@@ -0,0 +1,127 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/liqotech/liqo/pkg/consts"
+	"github.com/liqotech/liqo/pkg/utils"
+)
+
+var nodeNameField = "spec.nodeName"
+
+// NodeFailureReconciler reconciles a Node object
+type NodeFailureReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;delete
+
+// Reconcile nodes objects.
+func (r *NodeFailureReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Fetch the Node instance
+	var node corev1.Node
+	if err := r.Get(ctx, client.ObjectKey{Name: req.Name}, &node); err != nil {
+		klog.Errorf("an error occurred while getting node %s: %v", req.Name, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If node is Ready exit without doing anything
+	if utils.IsNodeReady(&node) {
+		return ctrl.Result{}, nil
+	}
+
+	// Node NotReady: delete all terminating pods that are managed by shadowpods
+	var pods corev1.PodList
+	labelSelector := client.MatchingLabelsSelector{Selector: labels.Set{consts.ManagedByLabelKey: consts.ManagedByShadowPodValue}.AsSelector()}
+	fieldSelector := client.MatchingFieldsSelector{Selector: fields.OneTermEqualSelector(nodeNameField, node.Name)}
+	if err := r.List(ctx, &pods, labelSelector, fieldSelector); err != nil {
+		klog.Errorf("unable to list pods: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	for _, pod := range pods.Items {
+		if !pod.DeletionTimestamp.IsZero() {
+			if err := r.Delete(ctx, &pod, client.GracePeriodSeconds(0)); err != nil {
+				klog.Errorf("unable to delete pod %q: %v", klog.KObj(&pod), err)
+				return ctrl.Result{}, err
+			}
+			klog.Infof("pod %q running on failed node %s deleted", klog.KObj(&pod), node.Name)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// getPodTerminatingEventHandler returns an event handler that reacts on Pod updates.
+// In particular, it reacts on changes over pods that are terminating and managed by a ShadowPod,
+// triggering the reconciliation of the related node hosting the pod.
+func getPodTerminatingEventHandler(c client.Client) handler.EventHandler {
+	return &handler.Funcs{
+		UpdateFunc: func(ue event.UpdateEvent, rli workqueue.RateLimitingInterface) {
+			ownedByShadowPod := ue.ObjectNew.GetLabels()[consts.ManagedByLabelKey] == consts.ManagedByShadowPodValue
+			isTerminating := !ue.ObjectNew.GetDeletionTimestamp().IsZero()
+			if ownedByShadowPod && isTerminating {
+				ctx := context.Background()
+				pod, ok := ue.ObjectNew.(*corev1.Pod)
+				if !ok {
+					klog.Errorf("object %v is not a pod", ue.ObjectNew)
+					return
+				}
+				nodeName := pod.Spec.NodeName
+				var node corev1.Node
+				if err := c.Get(ctx, client.ObjectKey{Name: nodeName}, &node); err != nil {
+					klog.Errorf("unable to get node %s (through pod %q): %v", nodeName, klog.KObj(pod), err)
+					return
+				}
```

Comment:
This is useless. You should not use the client in the handler function, and you don't need to

You may check that the `nodeName != ""` before add it